File: templates/code-carl/ch03_hashtable/t3_lc0349_intersection.py
"""LeetCode 349. Intersection of Two Arrays
https://leetcode.cn/problems/intersection-of-two-arrays/
"""

from typing import List


class Solution:
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        pass


if __name__ == "__main__":
    solution = Solution()
    assert sorted(solution.intersection([1, 2, 2, 1], [2, 2])) == [2]
    assert sorted(solution.intersection([4, 9, 5], [9, 4, 9, 8, 4])) == [4, 9]
    print("All tests passed.")

File: solutions/java/code-carl/ch03_hashtable/T3_LC0349_Intersection.java
/**
 * Practice template for:
 * - LeetCode 349. Intersection of Two Arrays
 *   https://leetcode.cn/problems/intersection-of-two-arrays/
 */
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class T3_LC0349_Intersection {
    public int[] intersection(int[] nums1, int[] nums2) {
        Set<Integer> set = new HashSet<>();
        List<Integer> res = new ArrayList<>();
        for (int n1 : nums1) {
            set.add(n1);
        }
        for (int n2 : nums2) {
            if (set.contains(n2)) {
                set.remove(n2);
                res.add(n2);
            }
        }
        return res.stream().mapToInt(Integer::intValue).toArray();
    }
}
